### PR TITLE
 Qt5 Updates

### PR DESCRIPTION
**Pull request by ar664**
This is what I have so far in terms of updating qt5. Any help would be appreciated.

### CODE REVIEW

**Comment by xtkoba** [comment]
Current context of build failure:

```
In file included from ../../corelib/kernel/qsharedmemory.cpp:41:
../../corelib/kernel/qsharedmemory_p.h:76:12: fatal error: 'sys/sem.h' file not found
#  include <sys/sem.h>
           ^~~~~~~~~~~
1 error generated.
```

I suggest including the following patch named `src-corelib-kernel-qsharedmemory_p.h.patch`:

```patch
--- a/src/corelib/kernel/qsharedmemory_p.h
+++ b/src/corelib/kernel/qsharedmemory_p.h
@@ -72,7 +72,7 @@
 # include "private/qobject_p.h"
 #endif
 
-#if !defined(Q_OS_WIN) && !defined(Q_OS_ANDROID) && !defined(Q_OS_INTEGRITY) && !defined(Q_OS_RTEMS)
+#if !defined(Q_OS_WIN) && !defined(Q_OS_ANDROID) && !defined(__ANDROID__) && !defined(Q_OS_INTEGRITY) && !defined(Q_OS_RTEMS)
 #  include <sys/sem.h>
 #endif
 
```

**Comment by ar664** [comment]
Ummm. I think the build failure was due to internet connectivity?

**Comment by xtkoba** [comment]
It was due to outdated submodule. Should be fixed now. Please rebase.

**Comment by ar664** [comment]
😯It built. Did I do everything correct?

**Comment by Grimler91** [comment]
@ar664 looks good to me, nice work!

Could you squash the commits, and/or start the commit message(s) with a prefix like: 
```
qt5-qt5base: 
```

(I would squash into a single commit with message `qt5-qt5base: update to 5.15.3`)

**Comment by ar664** [comment]
Ok, I'll fix the typo I made as well in build.sh

File: packages/qt5-qtbase/build.sh [hunk 1]
```diff
@@ -2,10 +2,10 @@ TERMUX_PKG_HOMEPAGE=https://www.qt.io/
 TERMUX_PKG_DESCRIPTION="A cross-platform application and UI framework"
 TERMUX_PKG_LICENSE="LGPL-3.0"
 TERMUX_PKG_MAINTAINER="Simeon Huang <symeon@librehat.com>"
-TERMUX_PKG_VERSION=5.12.11
-TERMUX_PKG_REVISION=11
-TERMUX_PKG_SRCURL="https://download.qt.io/official_releases/qt/5.12/${TERMUX_PKG_VERSION}/submodules/qtbase-everywhere-src-${TERMUX_PKG_VERSION}.tar.xz"
-TERMUX_PKG_SHA256=1c1b4e33137ca77881074c140d54c3c9747e845a31338cfe8680f171f0bc3a39
+TERMUX_PKG_VERSION=5.15.3
+TERMUX_PKG_REVISION=12
```
Comment (xtkoba) [inline packages/qt5-qtbase/build.sh:6]: Please reset (remove) `TERMUX_PKG_REVISION` when bumping version.

Comment (ar664) [reply packages/qt5-qtbase/build.sh:6]: Do you want me to make it 1 or remove the line?

Comment (xtkoba) [reply packages/qt5-qtbase/build.sh:6]: We normally remove the line, though equivalent if set to `0`.

**Comment by Yisus7u7** [comment]
@ar664 You need to rebuild all the apps that work with Qt5...

**Comment by xtkoba** [comment]
@Yisus7u7 Is it needed for just a minor bump without SOVERSION change? Yeah, there could be hidden API breakage, but we need evidence for that.

**Comment by ar664** [comment]
> @ar664 You need to rebuild all the apps that work with Qt5...

You got a list? 😗

**Comment by xtkoba** [comment]
https://wiki.qt.io/Qt-Version-Compatibility says

> Minor releases are backwards binary and source compatible.

I do not believe we have to rebuild reverse dependencies.

**Review by xtkoba — APPROVED** [review]

File: packages/qt5-qtbase/build.sh [hunk 2]
```diff
@@ -101,20 +100,20 @@ termux_step_configure () {
...
         -no-kms \
         -no-linuxfb \
-        -no-mirclient \
         -no-libudev \
         -no-evdev \
         -no-libinput \
         -no-mtdev \
         -no-tslib \
-        -system-xcb \
-        -no-xcb-xinput \
```
Comment (xtkoba) [inline packages/qt5-qtbase/build.sh:111]: `-no-xcb-xinput` is renamed to `-no-bundled-xcb-xinput` in Qt 5.15.

Without this option I expect some weird behavior described in commit 171e655d1fd6a3eeacffde46fb7a46fcd6be233b.

File: packages/qt5-qtbase/termux-build-qmake.sh [hunk 3]
```diff
@@ -63,14 +63,13 @@ termux_step_configure () {
...
         -no-kms \
         -no-linuxfb \
-        -no-mirclient \
         -no-libudev \
         -no-evdev \
         -no-libinput \
         -no-mtdev \
         -no-tslib \
-        -system-xcb \
-        -no-xcb-xinput \
```
Comment (xtkoba) [inline packages/qt5-qtbase/termux-build-qmake.sh:73]: Same here. I suppose `-no-bundled-xcb-xinput` is needed.

**Review by xtkoba — COMMENTED** [review]
The package name is `qt5-qtbase`, not qt5-qt5base.

**Review by xtkoba — COMMENTED** [review]
I wonder if other `qt5-*`s do not need to bump simultaneously:

* `qt5-qtdeclarative`
* `qt5-qtlocation`
* `qt5-qtmultimedia`
* `qt5-qtquickcontrols`
* `qt5-qtquickcontrols2`
* `qt5-qtsensors`
* `qt5-qtsvg`
* `qt5-qttools`
* `qt5-qtwebchannel`
* `qt5-qtwebsockets`
* `qt5-qtx11extras`
* `qt5-qtxmlpatterns`

File: packages/qt5-qtbase/build.sh (hunk 2 above)
Comment (xtkoba) [reply packages/qt5-qtbase/build.sh:111]: I'm confused. Should this rather be `-bundled-xcb-xinput`, which as an effect disables `system-xcb-xinput`?

It seems that we have to enable either `bundled-xcb-xinput` or `system-xcb-xinput` but cannot disable both of them.

File: packages/qt5-qtbase/termux-build-qmake.sh (hunk 3 above)
Comment (xtkoba) [reply packages/qt5-qtbase/termux-build-qmake.sh:73]: Or rather `-bundled-xcb-xinput`, maybe. (I'm not sure. See my comment above.)

**Comment by Yisus7u7** [comment]
> I wonder if other `qt5-*`s do not need to bump simultaneously:
> 
> * `qt5-qtdeclarative`
> * `qt5-qtlocation`
> * `qt5-qtmultimedia`
> * `qt5-qtquickcontrols`
> * `qt5-qtquickcontrols2`
> * `qt5-qtsensors`
> * `qt5-qtsvg`
> * `qt5-qttools`
> * `qt5-qtwebchannel`
> * `qt5-qtwebsockets`
> * `qt5-qtx11extras`
> * `qt5-qtxmlpatterns`

I'll try it

**Comment by Yisus7u7** [comment]
Screenshot_20220312-091216

@xtkoba If we will need to rebuild the apps written in Qt.

**Comment by ar664** [comment]
@xtkoba I'll start working on the build confs. 
@Yisus7u7 Guess we gotta do all the qt5-* at once.

File: packages/qt5-qtbase/build.sh (hunk 2 above)
Comment (ar664) [reply packages/qt5-qtbase/build.sh:111]: > `-no-xcb-xinput` is renamed to `-no-bundled-xcb-xinput` in Qt 5.15.
> 
> Without this option I expect some weird behavior described in commit [171e655](https://github.com/termux/x11-packages/commit/171e655d1fd6a3eeacffde46fb7a46fcd6be233b).

I guess we have to test it for Qt 5.15?

**Comment by Yisus7u7** [comment]
> @xtkoba I'll start working on the build confs. @Yisus7u7 Guess we gotta do all the qt5-* at once.

Yes.

After that, you have to rebuild the apps.

**Comment by xtkoba** [comment]
Current context of build failure in `qt5-qtdeclarative`:

```
/home/builder/.termux-build/qt5-qtdeclarative/src/bin/qmltyperegistrar --generate-qmltypes=plugins.qmltypes --import-name=QtQml --major-version=2 --minor-version=15 --foreign-types=/data/data/com.termux/files/usr/lib/metatypes/qt5network_metatypes.json,/data/data/com.termux/files/usr/lib/metatypes/qt5core_metatypes.json --dependencies=/home/builder/.termux-build/qt5-qtdeclarative/src/src/qml/dependencies.json --private-includes -o qt5qml_qmltyperegistrations.cpp qt5qml_metatypes.json
/home/builder/.termux-build/qt5-qtdeclarative/src/bin/qmltyperegistrar: 1: cannot open ]^E: No such file
```

The build system is trying to execute `qmltyperegistrar` binary built for Android target, which fails of course for cross building.

**Comment by xtkoba** [comment]
I wonder if it is just that `qmltyperegistrar` needs to be added in "Qt Declarative utilities":

https://github.com/termux/x11-packages/blob/c3dd4344ab588c1e08d270d4b9aed5c73e7e5869/packages/qt5-qtdeclarative/build.sh#L49-L50

https://github.com/termux/x11-packages/blob/c3dd4344ab588c1e08d270d4b9aed5c73e7e5869/packages/qt5-qtdeclarative/build.sh#L96-L97

**Comment by ar664** [comment]
> I wonder if it is just that `qmltyperegistrar` needs to be added in "Qt Declarative utilities":
> 
Well let's see. After I rebase and stuff.

**Comment by ar664** [comment]
The build checks are only there during the ready for review stage. Is it possible for a sample x86_64 build to be done in the draft stage? So we could waste less resources and use the draft feature.

**Comment by ar664** [comment]
I'm trying to build on my phone, but I get this error:
```
> aarch64-linux-android-clang++ -landroid-shmem -L/data/data/com.termux/files/usr/lib -fopenmp -static-openmp -Wl,--enable-new-dtags -Wl,--as-needed -Wl,-z,relro,-z,now -o arch arch.o  -L/data/data/com.termux/files/usr/lib 
Detected architecture: arm64 (neon)
test config.qtbase.tests.architecture gave result arm64
executing config test host_architecture 
+ cd /data/data/com.termux/files/home/.termux-build/qt5-qtbase/src/config.tests/arch && /data/data/com.termux/files/home/.termux-build/qt5-qtbase/src/bin/qmake "CONFIG -= qt debug_and_release app_bundle lib_bundle" "CONFIG += shared warn_off console single_arch" -early "CONFIG += cross_compile" /data/data/com.termux/files/home/.termux-build/qt5-qtbase/src/config.tests/arch/arch_host.pro 
> Project ERROR: failed to parse default search paths from compiler output
```

**Comment by xtkoba** [comment]
Yeah on device building can have its own difficulties.

**Comment by ar664** [comment]
> @ar664 And if we update only to qt5-qtbase?, and then we are updating little by little the other modules.
> 
> I don't know if you can, but qt5-qtbase is the most important package to upgrade

So disable the other packages in the mean time? Because I thought there would be significant changes in api between Major versions.

**Comment by Yisus7u7** [comment]
> > @ar664 And if we update only to qt5-qtbase?, and then we are updating little by little the other modules.
> > I don't know if you can, but qt5-qtbase is the most important package to upgrade
> 
> So disable the other packages in the mean time? Because I thought there would be significant changes in api between Major versions.

Must try

**Comment by xtkoba** [comment]
Ask the maintainer if it is allowed to update partially, if you are going to do so. I cannot decide it on my own.

**Comment by ar664** [comment]
> Ask the maintainer if it is allowed to update partially, if you are going to do so. I cannot decide it on my own.

@librehat

File: packages/qt5-qtlocation/build.sh [hunk 4]
```diff
@@ -2,10 +2,9 @@ TERMUX_PKG_HOMEPAGE=https://www.qt.io/
 TERMUX_PKG_DESCRIPTION="Qt 5 Location Library"
 TERMUX_PKG_LICENSE="LGPL-3.0"
 TERMUX_PKG_MAINTAINER="Simeon Huang <symeon@librehat.com>"
-TERMUX_PKG_VERSION=5.12.11
-TERMUX_PKG_REVISION=3
-TERMUX_PKG_SRCURL="https://download.qt.io/official_releases/qt/5.12/${TERMUX_PKG_VERSION}/submodules/qtlocation-everywhere-src-${TERMUX_PKG_VERSION}.tar.xz"
-TERMUX_PKG_SHA256=a3384357703f76fa2e352e4dfaa30d197098b77a3672ea85bd8925dbe357f00b
+TERMUX_PKG_VERSION=5.15.3
+TERMUX_PKG_SRCURL="https://download.qt.io/official_releases/qt/5.15/${TERMUX_PKG_VERSION}/submodules/qtlocation-everywhere-opensource-src-${TERMUX_PKG_VERSION}.tar.xz"
+TERMUX_PKG_SHA256=7889c9b4e54b2f9970d061201c064f04c2ed6d1f2b411bd93566958205fa4be7
```
Comment (xtkoba) [inline packages/qt5-qtlocation/build.sh:7]: The checksum for `qt5-qtlocation` seems to be incorrect (or outdated).

**Comment by xtkoba** [comment]
Please include https://github.com/termux/x11-packages/pull/643/commits/9876486b57e5a0fd8b24acb442ffecbc65c372dd into the PR. This should fix the build of `qt5-qtdeclarative`.

**Comment by librehat** [comment]
Unfortunately all Qt submodules must be updated at once (and then all the Qt applications need to be rebuilt) 
Thanks for your efforts on this. I haven't had time checking this myself. I wrote down this post when I was porting Qt 5.12:  https://www.librehat.com/building-qt-for-termux-android/ and it might be helpful if you encounter some issues.

Good luck!

**Comment by xtkoba** [comment]
Also include these changes:

* https://github.com/termux/x11-packages/pull/643/commits/9a99b9698db295fe1ce3458849cd9b40e647a3de
* https://github.com/termux/x11-packages/pull/643/commits/03478e4706cea4775efb3df69ce691b8f949e51f
* https://github.com/termux/x11-packages/pull/643/commits/1507df1793ea86e6299feafa65632203753ce1a9

**Comment by xtkoba** [comment]
OK, it builds. Then you should inspect the artifacts to see if anything is wrong with the contents.

I have already found some flaws in them. Will list patches later.

**Comment by ar664** [comment]
> OK, it builds. Then you should inspect the artifacts to see if anything is wrong with the contents.
> 
> I have already found some flaws in them. Will list patches later.

Should I rebase now or wait for patches?

**Comment by Yisus7u7** [comment]
> OK, it builds. Then you should inspect the artifacts to see if anything is wrong with the contents.
> 
> I have already found some flaws in them. Will list patches later.

I think it's already possible to merge the changes, right?

**Comment by xtkoba** [comment]
At your service.

For reference, here are the patches:

* https://github.com/termux/x11-packages/pull/643/commits/51675ef36d76e46e34fe06021def5edb880b28ff
* https://github.com/termux/x11-packages/pull/643/commits/72094bb420871b5ce3a28ad201afd9d30f2a2586

Note that they are not yet tested. It may fix the flaws or may break something.

**Comment by Yisus7u7** [comment]
> At your service.
> 
> For reference, here are the patches:
> 
> * [51675ef](https://github.com/termux/x11-packages/commit/51675ef36d76e46e34fe06021def5edb880b28ff)
> * [72094bb](https://github.com/termux/x11-packages/commit/72094bb420871b5ce3a28ad201afd9d30f2a2586)
> 
> Note that they are not yet tested. It may fix the flaws or may break something.

I will install the packages and compile lxqt, then tell you if I got any errors

**Comment by xtkoba** [comment]
Tests are always welcome. Thanks.

**Comment by xtkoba** [comment]
OK, please include:

* https://github.com/termux/x11-packages/pull/643/commits/51675ef36d76e46e34fe06021def5edb880b28ff
* https://github.com/termux/x11-packages/pull/643/commits/72094bb420871b5ce3a28ad201afd9d30f2a2586

**Comment by ar664** [comment]
Are there any other packages that need to be rebuilt?

**Comment by xtkoba** [comment]
If there is any, then it should be rebuilt in another PR.

I believe this PR is almost ready to merge. Please cleanup the commits. I suppose it is OK to squash them into one.

**Comment by ar664** [comment]
Ok, 🤞

**Comment by Yisus7u7** [comment]
Screenshot_20220318-145106

@xtkoba @ar664

I compiled all lxqt smoothly, you can merge the changes. 

I will take care of the update of lxqt

**Comment by Yisus7u7** [comment]
@xtkoba

**Pull request merged**